File: Simplex.kotlin/src/main/kotlin/kr/co/tmon/simplex/store/DisposableStore.kt
package kr.co.tmon.simplex.store

import kr.co.tmon.simplex.actions.*
import kr.co.tmon.simplex.channels.IChannel
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import java.util.concurrent.ConcurrentHashMap

class DisposableStore<TActionSet>(private val store: IActionStore<TActionSet>, private val subscriberId: String) : IActionStore<TActionSet> {

    private val managedDisposables: ConcurrentHashMap<String, CompositeDisposable> = ConcurrentHashMap()

    override fun <TAction : IAction<TParam, TResult>, TParam, TResult> dispatch(
        action: Function1<TActionSet, IActionBinder<TAction, TParam, TResult>>,
        parameters: TParam?,
        begin: Action?,
        end: Action?,
        channel: Function1<TAction, IChannel>? ) {
        store.dispatch(action, parameters, begin, end, channel)
    }

    override fun <TAction : IAction<TParam, TResult>, TParam, TResult> subscribe(
        action: Function1<TActionSet, IActionBinder<TAction, TParam, TResult>>,
        onNext: (TResult) -> Unit,
        observeOnMainThread: Boolean,
        observable: Function1<Observable<TResult>, Observable<TResult>>?,
        channel: Function1<TAction, IChannel>?,
        preventClone: Boolean
    ): Disposable? {
        return addDisposable(store.subscribe(action, onNext, observeOnMainThread, observable, channel, preventClone))
    }

    /**
     * 구독이 등록될때 발생하는 삭제패턴(Disposable)객체를 관리 목록에 등록합니다.
     * 관리목록은 모든 관리형 스토어에 공유되며, 삭제 패턴객체는 관리 스토어 식별자 별로 관리됩니다.
     * @param disposable 삭제패턴 객체
     * @return 삭제패턴 객체
     */
    private fun addDisposable(disposable: Disposable?) : Disposable? {
        if (disposable == null)
            return null

        val list = managedDisposables.get(subscriberId)
        if (list != null) {
            if (!list.isDisposed) {
                list.add(disposable)
            }
        } else {
            managedDisposables.put(subscriberId, CompositeDisposable(disposable))
        }

        return disposable
    }

    fun dispose() {
        //관리형 스토어별로 등록된 구독을 모두 해제 합니다.
        val value = managedDisposables.remove(subscriberId)
        if (value != null) {
            if (!value.isDisposed)
                value.dispose()
        }
    }
}

File: Simplex.Core/Store/DisposableStore.cs
using System;
using System.Collections.Concurrent;
using System.Reactive;
using System.Reactive.Disposables;
using System.Threading;
using Tmon.Simplex.Actions;
using Tmon.Simplex.Channels;

namespace Tmon.Simplex.Store
{
    public class DisposableStore<TActionBinderSet> : IActionStore<TActionBinderSet>, IDisposable
        where TActionBinderSet : class, IActionBinderSet, new()
    {
        private IActionStore<TActionBinderSet> Store { get; }

        private readonly string subscriberId;

        private static readonly ConcurrentDictionary<string, CompositeDisposable> managedDisposables 
            = new ConcurrentDictionary<string, CompositeDisposable>();

        private DisposableStore() { }

        /// <summary>
        /// 스토어에 식별자를 부여하여 관리되는 스토어를 생성합니다.
        /// </summary>
        /// <param name="store">스토어 객체</param>
        /// <param name="subscriberId">부여할 식별자</param>
        public DisposableStore(IActionStore<TActionBinderSet> store, string subscriberId)
        {
            this.Store = store;
            this.subscriberId = subscriberId;
        }

        public void Dispatch<TAction, TResult>(
            Func<TActionBinderSet, IActionBinder<TAction, TResult>> action,
            Action begin = null,
            Action end = null,
            Func<TAction, Channel> channel = null,
            CancellationToken? cancellationToken = null)
            where TAction : IAction<TResult>
            => Store.Dispatch(action, begin, end, channel, cancellationToken);

        public void Dispatch<TAction, TParam, TResult>(
            Func<TActionBinderSet, IActionBinder<TAction, TParam, TResult>> action,
            TParam parameter,
            Action begin = null,
            Action end = null,
            Func<TAction, Channel> channel = null,
            CancellationToken? cancellationToken = null)
            where TAction : IAction<TParam, TResult>
            => Store.Dispatch(action, parameter, begin, end, channel, cancellationToken);

        public IDisposable Subscribe<TAction>(
            Func<TActionBinderSet, IActionBinder<TAction, Unit>> action,
            Action onNext,
            bool observeOnMainThread = false,
            Func<IObservable<Unit>, IObservable<Unit>> observable = null,
            Func<TAction, Channel> channel = null)
            where TAction : IAction<Unit>
            => AddDisposable(Store.Subscribe(action, onNext, observeOnMainThread, observable, channel));
       
        public IDisposable Subscribe<TAction, TResult>(
            Func<TActionBinderSet, IActionBinder<TAction, TResult>> action,
            Action<TResult> onNext,
            bool observeOnMainThread = false,
            Func<IObservable<TResult>, IObservable<TResult>> observable = null,
            Func<TAction, Channel> channel = null,
            bool preventClone = false)
            where TAction : IAction<TResult>
            => AddDisposable(Store.Subscribe(action, onNext, observeOnMainThread, observable, channel, preventClone));
        
        /// <summary>
        /// 구독이 등록될때 발생하는 삭제패턴(IDisposable)객체를 관리 목록에 등록합니다.
        /// 관리목록은 모든 관리형 스토어에 공유되며, 삭제 패턴객체는 관리 스토어 식별자 별로 관리됩니다.
        /// </summary>
        /// <param name="disposables">삭제패턴 객체</param>
        /// <returns>삭제패턴 객체</returns>
        private IDisposable AddDisposable(IDisposable disposable)
        {
            if (disposable == null)
                return null;

            managedDisposables.AddOrUpdate(
                key: subscriberId,
                addValueFactory: key =>
                {
                    return new CompositeDisposable(disposable);
                },
                updateValueFactory: (key, list) =>
                {
                    if (!list.IsDisposed)
                    {
                        list.Add(disposable);
                    }
                    return list;
                });
            
            return disposable;
        }
        
        public void Dispose()
        {
            //관리형 스토어별로 등록된 구독을 모두 해제 합니다.
            if (managedDisposables.TryRemove(subscriberId, out CompositeDisposable value) && value != null)
            {
                if (!value.IsDisposed)
                    value.Dispose();
            }
        }
    }
}
